File: packages/cli/src/templates/_shared/lib/context/task_tracker.py
"""Task tracker — direct state.json CRUD for tasks.

Writes tasks directly to the tasks[] array in state.json,
bypassing events.jsonl for faster, simpler task operations.

All functions do their own I/O to avoid circular imports with
context_store.py.
"""
import json
import re
from pathlib import Path
from typing import Dict, List, Optional

from ..base.atomic_write import atomic_write
from ..base.constants import get_context_dir
from ..base.utils import eprint, now_iso


# ---------------------------------------------------------------------------
# Internal I/O (avoids circular import with context_store)
# ---------------------------------------------------------------------------

def _state_path(context_id: str, project_root: Path = None) -> Path:
    return get_context_dir(context_id, project_root) / "state.json"


def _load_state(context_id: str, project_root: Path = None) -> Optional[dict]:
    sp = _state_path(context_id, project_root)
    if not sp.exists():
        return None
    try:
        return json.loads(sp.read_text(encoding="utf-8"))
    except Exception as e:
        eprint(f"[task_tracker] WARNING: Failed to read state.json: {e}")
        return None


def _save_state(context_id: str, state_data: dict, project_root: Path = None) -> bool:
    sp = _state_path(context_id, project_root)
    content = json.dumps(state_data, indent=2, ensure_ascii=False)
    success, error = atomic_write(sp, content)
    if not success:
        eprint(f"[task_tracker] WARNING: Failed to write state.json: {error}")
    return success


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def generate_next_task_id(context_id: str, project_root: Path = None) -> str:
    """Scan tasks[] for highest aiw-N, return aiw-(N+1)."""
    state = _load_state(context_id, project_root)
    tasks = state.get("tasks", []) if state else []

    max_num = 0
    for t in tasks:
        tid = t.get("id", "")
        m = re.match(r"^aiw-(\d+)$", tid)
        if m:
            max_num = max(max_num, int(m.group(1)))

    return f"aiw-{max_num + 1}"


def add_task(
    context_id: str,
    subject: str,
    description: str = "",
    active_form: str = "",
    session_id: str = "",
    project_root: Path = None,
) -> Optional[dict]:
    """Add a new task to state.json tasks[] and return the task dict."""
    state = _load_state(context_id, project_root)
    if state is None:
        return None

    task_id = generate_next_task_id(context_id, project_root)
    task = {
        "id": task_id,
        "subject": subject,
        "description": description,
        "active_form": active_form,
        "status": "pending",
        "created_at": now_iso(),
        "completed_at": None,
        "evidence": "",
        "work_summary": "",
        "files_changed": [],
        "session_id": session_id,
    }

    state.setdefault("tasks", []).append(task)
    state["last_active"] = now_iso()

    if _save_state(context_id, state, project_root):
        return task
    return None


def update_task(
    context_id: str,
    task_id: str,
    status: str = None,
    evidence: str = "",
    work_summary: str = "",
    files_changed: List[str] = None,
    session_id: str = "",
    project_root: Path = None,
) -> bool:
    """Find task by task_id in tasks[], update fields, return True on success."""
    state = _load_state(context_id, project_root)
    if state is None:
        return False

    for task in state.get("tasks", []):
        if task.get("id") == task_id:
            if status is not None:
                task["status"] = status
                if status == "completed":
                    task["completed_at"] = now_iso()
            if evidence:
                task["evidence"] = evidence
            if work_summary:
                task["work_summary"] = work_summary
            if files_changed is not None:
                task["files_changed"] = files_changed
            if session_id:
                task["session_id"] = session_id
            state["last_active"] = now_iso()
            return _save_state(context_id, state, project_root)

    eprint(f"[task_tracker] Task '{task_id}' not found in context '{context_id}'")
    return False


def delete_task(context_id: str, task_id: str, project_root: Path = None) -> bool:
    """Remove task from tasks[] and return True on success."""
    state = _load_state(context_id, project_root)
    if state is None:
        return False

    tasks = state.get("tasks", [])
    original_len = len(tasks)
    state["tasks"] = [t for t in tasks if t.get("id") != task_id]

    if len(state["tasks"]) == original_len:
        eprint(f"[task_tracker] Task '{task_id}' not found in context '{context_id}'")
        return False

    state["last_active"] = now_iso()
    return _save_state(context_id, state, project_root)


def get_tasks(context_id: str, project_root: Path = None) -> List[dict]:
    """Return tasks[] from state.json."""
    state = _load_state(context_id, project_root)
    if state is None:
        return []
    return state.get("tasks", [])


def generate_task_summary(context_id: str, project_root: Path = None) -> str:
    """Partition tasks and format as markdown checklist."""
    tasks = get_tasks(context_id, project_root)
    if not tasks:
        return "No tasks in this context."

    completed = [t for t in tasks if t.get("status") == "completed"]
    in_progress = [t for t in tasks if t.get("status") == "in_progress"]
    pending = [t for t in tasks if t.get("status") == "pending"]
    blocked = [t for t in tasks if t.get("status") == "blocked"]

    lines = [f"### Tasks ({len(tasks)} total)", ""]

    for t in completed:
        ws = f"\n  Work: {t['work_summary']}" if t.get("work_summary") else ""
        lines.append(f"- [x] {t['id']}: {t['subject']}{ws}")
    for t in in_progress:
        lines.append(f"- [~] {t['id']}: {t['subject']}")
    for t in pending:
        lines.append(f"- [ ] {t['id']}: {t['subject']}")
    for t in blocked:
        lines.append(f"- [!] {t['id']}: {t['subject']}")

    return "\n".join(lines)

File: packages/cli/src/templates/core/lib-ts/context/task-tracker.ts
/**
 * Task tracker — direct state.json CRUD for tasks.
 * See SPEC.md §10
 *
 * Writes tasks directly to the tasks[] array in state.json.
 * Uses state-io for I/O to avoid circular imports with context-store.
 */

import { logWarn } from "../runtime/logger.js";
import { readStateJson, toDict as _toDict, writeStateJson } from "../runtime/state-io.js";
import { nowIso } from "../runtime/utils.js";
import type { ContextState as _ContextState, Task } from "../types.js";

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Scan tasks[] for highest aiw-N, return aiw-(N+1).
 * See SPEC.md §10.2
 */
export function generateNextTaskId(contextId: string, projectRoot?: string): string {
  const state = readStateJson(contextId, projectRoot);
  const tasks = state?.tasks ?? [];

  let maxNum = 0;
  for (const t of tasks) {
    const match = /^aiw-(\d+)$/.exec(t.id);
    if (match) {
      const num = Number.parseInt(match[1]!, 10);
      if (num > maxNum) maxNum = num;
    }
  }

  return `aiw-${maxNum + 1}`;
}

/**
 * Add a new task to state.json tasks[] and return the task object.
 * See SPEC.md §10.3
 */
export function addTask(
  contextId: string,
  subject: string,
  description = "",
  activeForm = "",
  sessionId = "",
  projectRoot?: string,
): null | Task {
  const state = readStateJson(contextId, projectRoot);
  if (!state) return null;

  const taskId = generateNextTaskId(contextId, projectRoot);
  const task: Task = {
    id: taskId,
    subject,
    description,
    active_form: activeForm,
    status: "pending",
    created_at: nowIso(),
    completed_at: null,
    evidence: "",
    work_summary: "",
    files_changed: [],
    session_id: sessionId,
  };

  state.tasks.push(task);
  state.last_active = nowIso();

  const [success] = writeStateJson(contextId, state, projectRoot);
  return success ? task : null;
}

/**
 * Find task by task_id in tasks[], update fields, return true on success.
 * See SPEC.md §10.4
 */
export function updateTask(
  contextId: string,
  taskId: string,
  opts?: {
    evidence?: string;
    files_changed?: string[];
    session_id?: string;
    status?: string;
    work_summary?: string;
  },
  projectRoot?: string,
): boolean {
  const state = readStateJson(contextId, projectRoot);
  if (!state) return false;

  for (const task of state.tasks) {
    if (task.id === taskId) {
      if (opts?.status !== undefined) {
        task.status = opts.status as Task["status"];
        if (opts.status === "completed") {
          task.completed_at = nowIso();
        }
      }

      if (opts?.evidence) task.evidence = opts.evidence;
      if (opts?.work_summary) task.work_summary = opts.work_summary;
      if (opts?.files_changed !== undefined) task.files_changed = opts.files_changed;
      if (opts?.session_id) task.session_id = opts.session_id;
      state.last_active = nowIso();
      const [success] = writeStateJson(contextId, state, projectRoot);
      return success;
    }
  }

  logWarn("task_tracker", `Task '${taskId}' not found in context '${contextId}'`);
  return false;
}

/**
 * Remove task from tasks[] and return true on success.
 * See SPEC.md §10.5
 */
export function deleteTask(
  contextId: string,
  taskId: string,
  projectRoot?: string,
): boolean {
  const state = readStateJson(contextId, projectRoot);
  if (!state) return false;

  const originalLen = state.tasks.length;
  state.tasks = state.tasks.filter(t => t.id !== taskId);

  if (state.tasks.length === originalLen) {
    logWarn("task_tracker", `Task '${taskId}' not found in context '${contextId}'`);
    return false;
  }

  state.last_active = nowIso();
  const [success] = writeStateJson(contextId, state, projectRoot);
  return success;
}

/**
 * Return tasks[] from state.json.
 * See SPEC.md §10.6
 */
export function getTasks(contextId: string, projectRoot?: string): Task[] {
  const state = readStateJson(contextId, projectRoot);
  if (!state) return [];
  return state.tasks;
}

/**
 * Partition tasks and format as markdown checklist.
 * See SPEC.md §10.7
 */
export function generateTaskSummary(contextId: string, projectRoot?: string): string {
  const tasks = getTasks(contextId, projectRoot);
  if (tasks.length === 0) return "No tasks in this context.";

  const completed = tasks.filter(t => t.status === "completed");
  const inProgress = tasks.filter(t => t.status === "in_progress");
  const pending = tasks.filter(t => t.status === "pending");
  const blocked = tasks.filter(t => t.status === "blocked");

  const lines: string[] = [`### Tasks (${tasks.length} total)`, ""];

  for (const t of completed) {
    const ws = t.work_summary ? `\n  Work: ${t.work_summary}` : "";
    lines.push(`- [x] ${t.id}: ${t.subject}${ws}`);
  }

  for (const t of inProgress) {
    lines.push(`- [~] ${t.id}: ${t.subject}`);
  }

  for (const t of pending) {
    lines.push(`- [ ] ${t.id}: ${t.subject}`);
  }

  for (const t of blocked) {
    lines.push(`- [!] ${t.id}: ${t.subject}`);
  }

  return lines.join("\n");
}
